Registry compatibility level

nucccc opened the pull request:
Wrote code for compatibility level putting and getting. For the testing a bit of guidance would be well accepted! I usually set my tests with a container running, but I couldn't manage to run the whole zookeper+kafka_broker+confluent_schema_registry using [testcontainers-go](https://github.com/testcontainers/testcontainers-go)

Thanks for the attention,
Nuc

nrwiersma commented:
Hi.

For tests, the standard is unit tests. An example can be found here: https://github.com/hamba/avro/blob/main/registry/client_test.go#L110.

You code need to pass the linter as well.

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+*/
+func validCompatibilityLevel(compatibilityLevel string) bool {
+	switch compatibilityLevel {
+	case BackwardCL:
+		return true
+	case BackwardTransitiveCL:
+		return true
+	case ForwardCL:
+		return true
+	case ForwardTransitiveCL:
+		return true
+	case FullCL:
+		return true
+	case FullTransitiveCL:
+		return true
+	case NoneCL:
+		return true
```
This all can be a single `case`.

nucccc replied:
didn't know the syntax thank you

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+https://docs.confluent.io/platform/current/schema-registry/develop/api.html#compatibility
+.
+*/
+func validCompatibilityLevel(compatibilityLevel string) bool {
```
It is more common for validation functions to return errors, and `compatibilityLevel` is very verbose, `lvl` is more than enough to know what it is:
```suggestion
func validateCompatibilityLevel(lvl string) error {
```

nucccc replied:
done

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+is valid according to the ones described in:
+https://docs.confluent.io/platform/current/schema-registry/develop/api.html#compatibility
+.
+*/
```
Block comments are basically never used in Go, this should be a set of line comments. Also sentences should start with a capital letter.

nucccc replied:
done

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+}
+
+// PutCompatibilityLevelGlobal sets the global compatibility level of the registry.
+func (c *Client) PutCompatibilityLevelGlobal(
```
`SetGlobalCompatibilityLevel` would fit better here.

nucccc replied:
done

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+// PutCompatibilityLevelGlobal sets the global compatibility level of the registry.
+func (c *Client) PutCompatibilityLevelGlobal(
+	ctx context.Context,
+	compatibilityLevel string,
```
`lvl` is enough here.

nucccc replied:
done

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+}
+
+// GetCompatibilityLevelGlobal gets the global compatibility level.
+func (c *Client) GetCompatibilityLevelGlobal(
```
`GetGlobalCompatibilityLevel` would be better.

nucccc replied:
done

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+func (c *Client) PutCompatibilityLevelGlobal(
+	ctx context.Context,
+	compatibilityLevel string,
+) (string, error) {
```
I am confused by the returned compatibility level (`string`) in this function and the next? Either it worked and it is the compatibility level you set, or it didn't and it would return an error. What is the need here?

nucccc replied:
The API returns the compatibility level that was set according to the [docs](GetCompatibilityLevelGlobal). So boh, I thought of returning it that well (shrug)

nucccc replied:
Regarding this point, personally for me it's the same.

I would say that a method returning the compatibility level could more "related" to the actual API, while a mehod not returning that string would make maybe more sense from the usage point of view.

I'm open to every decision you make regarding this

nrwiersma replied:
In general I think it would lead to a lot of code ignoring the return, and I cannot think of a single use cause where the return would be used. Personally I would not return it.

nucccc replied:
done

nucccc commented:
In the latest commit i made my best efforts to come towards the requests of the linter, and added some unit tests.

To me it appeared the linter required comments to end with a period, I didn't do that, obtaining an error. I also made my best to avoid warnings by setting the beginning of the comments as the name of the method/function.

I also added a unite test for each client method proposed, and a unit test for the function verifying if a compatibility level is valid according to confluent docs.

I hope the work makes sense, and I would kindly ask you if it is possible torerun the checks.

Thank you in advance,
Nuc

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+https://docs.confluent.io/platform/current/schema-registry/develop/api.html#compatibility
+.
+*/
+func ValidCompatibilityLevel(compatibilityLevel string) bool {
```
What is the reason to make this public? You already expose the levels.

nucccc replied:
Exclusively to call it in the unit tests

nucccc replied:
dumbness of mine, I moved it to client_internal_test.go

nrwiersma replied:
There is not need to expose it just for the test as it muddies the API of the package. It should be tested via the error tests in the functions themselves, like `TestClient_SetGlobalCompatibilityLevelHandlesInvalidLevel`.

nucccc replied:
Did my best to put also these tests

nrwiersma commented on registry/client.go:
```diff
@@ -352,3 +352,118 @@ func (e Error) Error() string {
+level not present in:
+https://docs.confluent.io/platform/current/schema-registry/develop/api.html#compatibility .
+*/
+func errByInvalidCL(compatibilityLevel string) error {
+	return fmt.Errorf("invalid compatibility level %s", compatibilityLevel)
+}
```
I find a function with a single line error very odd. I get the point of reuse, but this is odd none the less. It could perhaps be returned by the validation function instead.

nucccc replied:
removed

nrwiersma reviewed (changes requested):
Thanks for the contribution. I have added some comments. Additionally your tests need to cover the error cases of the functions as well.

nucccc commented:
I think I solved almost everything required.

I leave a point open:
- the functions setting the compatibility level actually return the string, complying with what the API does (shrug!). In my personal opinion you can decide what you prefer, it's up to you.

Nuc

nucccc commented:
Made my best to come towards all actual revisions

nrwiersma reviewed (approved):
LGTM 🎉 Thanks for the contribution.
